### PR TITLE
Fixed recently introduced notifications related null reference exceptions

### LINKED ISSUE #19837: NullReferenceException in OpenRA.Sound.PlayPredefined
Happened on https://github.com/OpenRA/OpenRA/commit/3bde4ebbaf9801e58c3f62a516670db015950bca after some play time where I annoyed the AI and it started to hunt my units down.

```
Exception of type `System.NullReferenceException`: Object reference not set to an instance of an object.
   at OpenRA.Sound.PlayPredefined(SoundType soundType, Ruleset ruleset, Player p, Actor voicedActor, String type, String definition, String variant, Boolean relative, WPos pos, Single volumeModifier, Boolean attenuateVolume) in /home/matthias/Entwicklung/OpenHV/engine/OpenRA.Game/Sound/Sound.cs:line 344
   at OpenRA.Sound.PlayNotification(Ruleset rules, Player player, String type, String notification, String variant) in /home/matthias/Entwicklung/OpenHV/engine/OpenRA.Game/Sound/Sound.cs:line 426
   at OpenRA.Mods.Common.Traits.Sound.ActorLostNotification.OpenRA.Mods.Common.Traits.INotifyKilled.Killed(Actor self, AttackInfo e) in /home/matthias/Entwicklung/OpenHV/engine/OpenRA.Mods.Common/Traits/Sound/ActorLostNotification.cs:line 38
   at OpenRA.Mods.Common.Traits.Health.InflictDamage(Actor self, Actor attacker, Damage damage, Boolean ignoreModifiers) in /home/matthias/Entwicklung/OpenHV/engine/OpenRA.Mods.Common/Traits/Health.cs:line 221
   at OpenRA.Mods.Common.Warheads.SpreadDamageWarhead.DoImpact(WPos pos, Actor firedBy, WarheadArgs args) in /home/matthias/Entwicklung/OpenHV/engine/OpenRA.Mods.Common/Warheads/SpreadDamageWarhead.cs:line 106
   at OpenRA.GameRules.WeaponInfo.Impact(Target& target, WarheadArgs args) in /home/matthias/Entwicklung/OpenHV/engine/OpenRA.Game/GameRules/WeaponInfo.cs:line 235
   at OpenRA.Mods.Common.Projectiles.Bullet.Explode(World world) in /home/matthias/Entwicklung/OpenHV/engine/OpenRA.Mods.Common/Projectiles/Bullet.cs:line 324
   at OpenRA.WorldUtils.DoTimed[T](IEnumerable`1 e, Action`1 a, String text) in /home/matthias/Entwicklung/OpenHV/engine/OpenRA.Game/WorldUtils.cs:line 75
   at OpenRA.World.Tick() in /home/matthias/Entwicklung/OpenHV/engine/OpenRA.Game/World.cs:line 430
   at OpenRA.Game.InnerLogicTick(OrderManager orderManager) in /home/matthias/Entwicklung/OpenHV/engine/OpenRA.Game/Game.cs:line 623
   at OpenRA.Game.LogicTick() in /home/matthias/Entwicklung/OpenHV/engine/OpenRA.Game/Game.cs:line 638
   at OpenRA.Game.Loop() in /home/matthias/Entwicklung/OpenHV/engine/OpenRA.Game/Game.cs:line 803
   at OpenRA.Game.Run() in /home/matthias/Entwicklung/OpenHV/engine/OpenRA.Game/Game.cs:line 856
   at OpenRA.Game.InitializeAndRun(String[] args) in /home/matthias/Entwicklung/OpenHV/engine/OpenRA.Game/Game.cs:line 294
   at OpenRA.Launcher.Program.Main(String[] args) in /home/matthias/Entwicklung/OpenHV/engine/OpenRA.Launcher/Program.cs:line 32
```

### PR DESCRIPTION
**Pull request by Mailaender**
Closes https://github.com/OpenRA/OpenRA/issues/19837

### CODE REVIEW

File: OpenRA.Mods.Common/Traits/Sound/ActorLostNotification.cs [hunk 1]
```diff
@@ -20,21 +20,29 @@ class ActorLostNotificationInfo : ConditionalTraitInfo
...
-			: base(info) { }
+		readonly World world;
+		readonly Player player;
+		readonly FactionInfo faction;
+
+		public ActorLostNotification(ActorLostNotificationInfo info, Actor self)
+			: base(info)
+		{
+			world = self.World;
+			player = Info.NotifyAll ? world.LocalPlayer : self.Owner;
```
Comment (reaperrr) [inline OpenRA.Mods.Common/Traits/Sound/ActorLostNotification.cs]: What if the actor is captured and therefore changes owner midgame?

**Comment by pchote** [comment]
I don't see how these changes could avoid any NREs inside `PlayPredefined`. What am I missing?

**Comment by Mailaender** [comment]
I believe the crash to be related to `self.World.Map.Rules` where `self` is disposed.

**Comment by pchote** [comment]
`INotifyKilled.Killed` is never called on a disposed actor (it is called as the last thing before disposing) - if this is somehow happening, then it needs to be fixed at the source!

**Comment by Mailaender** [comment]
It is clearly a regression from https://github.com/OpenRA/OpenRA/pull/19667 that I don't yet understand.

File: OpenRA.Mods.Common/Traits/Sound/ActorLostNotification.cs (hunk 1 above)
Comment (Mailaender) [reply OpenRA.Mods.Common/Traits/Sound/ActorLostNotification.cs]: This change is misleading. This trait is not always the source, so this won't help at all.

**Comment by Mailaender** [comment]
I found a fix. Not sure why the stacktraces are so misleading.

**Comment by RoosterDragon** [comment]
Is the issue here that a null sound ends up in `currentNotifications` which later hurts us? In that case, would it make more sense to add a check after `var sound = soundEngine.Play2D(...` (line 410) and do `if (sound == null) return false;`? That would tell the caller the sound failed to play, and stops us putting nulls into either `currentNotifications` or `currentSounds` which makes life easier to think about.

**Comment by Mailaender** [comment]
Added that as well. I don't know why the notifications are null. The rules seem to be correct.

**Comment by Dzierzan** [comment]
I am sorry to be an ass, but can someone review that? We wanted to release another pre-release version of OpenHV and we need this fix.

**Review by abcdefg30 — APPROVED** [review]
Untested but changes lgtm.

**Review by reaperrr — APPROVED** [review]

**Comment by reaperrr** [comment]
https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/8e9317262f3ec889028be03a11bf16b931977a53...d40247b1d27a8886766c6e9708741aba641110d1

**Pull request merged**